internal/Yesod/Session/Memcache/Storage.hs

```diff
@@ -0,0 +1,91 @@
+module Yesod.Session.Memcache.Storage
+  ( memcacheStorage
+  , SessionPersistence (..)
+  ) where
+
+import Internal.Prelude
+
+import Control.Monad.Reader (MonadReader (ask))
+import Database.Memcache.Client qualified as Memcache
+import Database.Memcache.Types qualified as Memcache
+import Session.Key
+import Yesod.Session.SessionType
+import Yesod.Session.Storage.Exceptions
+import Yesod.Session.Storage.Operation
+
+-- | Mapping between 'Session' and Memcache representation.
+data SessionPersistence = SessionPersistence
+  { databaseKey :: SessionKey -> Memcache.Key
+  , toDatabase :: Session -> Memcache.Value
+  , fromDatabase :: Memcache.Value -> Session
+  , runDB :: forall a b m. ReaderT b IO a -> m a
```

Comment:
Since for Memcache a lot of the Persistent abstractions aren't present and there's no connection pool, I think instead of a `runDB` function all you really want here is `, client :: Memcache.Client`.

Then in the `Yesod.Session.Memcache.Yesod` module:

```haskell
makeSessionBackend configuration =
  makeSessionBackend'
    SessionConfiguration'
      { storage = memcacheStorage configuration.persistence
      , options = configuration.options
      , runDB = flip runReaderT configuration.persistence.client
      }
```